rfd/0207-encrypted-session-recordings.md

```diff
@@ -0,0 +1,293 @@
+`--decryption-key-path` respectively. When playing recordings managed by an HSM, the
+identity will be wrapped and included as the first line of the recording which will
+be unwrapped using the same `--hsm*` parameters as `tctl keygen`.
+
+### HSMs
+
+The implementation described should be sufficient for integrating HSMs where
+human-in-the-loop is possible. Where human-in-the-loop is not possible, such as
+private keys that can not be exported after creation, we will need a more direct
+integration with the HSM. Teleport already integrates with PKCS#11, AWS KMS, and GCP
+KMS as keystore implementations. It should be possible to leverage these integrations
+to call necessary functions within supported HSMs.
+
+#### HSM Wrapped Data Key
+
+In order to facilitate encryption with the HSM, we will use the
+`tctl keygen --hsm-pin=12345` utility to generate an X25519 key pair much like the
+standard flow. The utility would then integrate with the HSM's key wrapping
+functionality to wrap the private key and combine it with the public key output such
+as:
+`<known-length-age-public-key><base64-hsm-wrapped-age-private-key>`
+The combined output could then be configured as a recipient just like an ordinary
+`age_x25519` public key. This method would require a small modification to the final
+`.tar` file to include the wrapped key such that `tsh play` would be able to find it
+and call the HSM in order to unwrap the private identity for decryption. This method
+allows the HSM to manage the secret key material without requiring it to decrypt
+potentially large session recordings. The HSM is also free to determine the key
+algorithm for wrapping without being constrained to what is supported by `age`.
+
+#### HSM Configuration
+
+HSM integration happens through PKCS#11 and is configured through two CLI parameters.
+
+The `--hsm-pin=<pin>` parameter is specific to the `pkcs11` provider and defines the pin
+value that should be used to authentice with the configured HSM. This is also the
+parameter that informs `tctl keygen` that it should generate an HSM wrapped key.
+
+The `--hsm-config-dir=<path>` parameter points to a PKCS#11 configuration directory.
+On unix hosts, the default path is `$XDG_CONFIG/pkcs11/`. It is expected to contain
+a `pkcs11.conf` file and a `modules/` directory containing configurations for PKCS#11
+modules such as the PKCS#11 library for AWS KMS CloudHSM Client SDK.
+
+
+### Key Rotation
+
+Key rotation is possible for new recordings by adding a new recipient entry to
+the `session_recording_recipients`. Because recordings captured can be retrieved
+using identities associated with any recipient used during encryption, it is possible
+to verify that new recordings are still retrievable before removing the rotated
+recipient. Reencrypting recordings captured before the rotation is not in scope for
+this implementation. A potential method for ensuring continued access to historical
```

Comment:
A possible option here could be to encrypt all recording with a long-term key associated to the first admin user created on Teleport. Having said that, if key rotation is not required for old recordings, I wouldn't increase the complexity of keys management

Reply:
I agree, and I don't think this is a requirement we need to support :+1: